File: Dynamic Programming/1931.Painting a Grid With Three Different Colors/Solution.py
class Solution:
    def colorTheGrid(self, m: int, n: int) -> int:
        def dfs(path):
            if len(path) == m:
                masks.append(tuple(path))
                return
            for k in range(3):
                if path and path[-1] == k:
                    continue
                path.append(k)
                dfs(path)
                path.pop()

        masks = []
        dfs([])
        g = defaultdict(set)
        for i, m1 in enumerate(masks):
            for j, m2 in enumerate(masks):
                if any(m1[k] == m2[k] for k in range(m)):
                    continue
                g[i].add(j)
                g[j].add(i)

        f = [[0] * len(masks) for _ in range(n)]
        for i in range(len(masks)):
            f[0][i] = 1
        for i in range(n - 1):
            for j in range(len(masks)):
                for k in g[j]:
                    f[i + 1][k] += f[i][j]
        return sum(f[-1]) % int(1e9 + 7)

File: Dynamic Programming/1931.Painting a Grid With Three Different Colors/Solution.java
class Solution {

    List<String> masks;

    public void dfs(List<String> path, int m) {
        if (path.size() == m) {
            masks.add(String.join("", path));
            return;
        }
        for (String k : new String[]{"0", "1", "2"}) {
            if (!path.isEmpty() && path.get(path.size() - 1).equals(k)) {
                continue;
            }
            path.add(k);
            dfs(path, m);
            path.remove(path.size() - 1);
        }
    }

    public int colorTheGrid(int m, int n) {
        masks = new ArrayList<>();
        dfs(new ArrayList<>(), m);

        Map<Integer, Set<Integer>> g = new HashMap<>();
        for (int i = 0; i < masks.size(); i++) {
            for (int j = i + 1; j < masks.size(); j++) {
                boolean diff = true;
                for (int k = 0; k < m; k++) {
                    if (masks.get(i).charAt(k) == masks.get(j).charAt(k)) {
                        diff = false;
                        break;
                    }
                }
                if (diff) {
                    g.computeIfAbsent(i, k -> new HashSet<>()).add(j);
                    g.computeIfAbsent(j, k -> new HashSet<>()).add(i);
                }
            }
        }

        int[][] f = new int[n][masks.size()];
        Arrays.fill(f[0], 1);
        int M = (int) (1e9 + 7);
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < masks.size(); j++) {
                for (int k : g.get(j)) {
                    f[i + 1][k] = (f[i + 1][k] + f[i][j]) % M;
                }
            }
        }
        int ans = 0;
        for (int i = 0; i < masks.size(); i++) {
            ans = (ans + f[n - 1][i]) % M;
        }
        return ans;
    }
}